File: src/main/java/me/oriharel/machinery/upgrades/AbstractUpgrade.java
package me.oriharel.machinery.upgrades;

import com.google.common.base.Preconditions;
import me.oriharel.machinery.Machinery;
import me.oriharel.machinery.machine.MachineResourceGetProcess;
import org.bukkit.configuration.file.YamlConfiguration;

import java.util.HashMap;
import java.util.Map;

/**
 * abstract implementation of upgrade.
 * abstract since every upgrade has it's own modifying logic but behind the scenes upgrading and downgrading a upgrade is the same.
 */
public abstract class AbstractUpgrade {
    protected int level;
    protected boolean runOnlyOnProcessStart;
    protected YamlConfiguration configLoad;
    protected int maxLevel = -1;

    public AbstractUpgrade(int level) {
        this.level = level;
        this.runOnlyOnProcessStart = false;
        this.configLoad = Machinery.getInstance().getFileManager().getConfig("upgrades.yml").get();
    }

    protected AbstractUpgrade() {
        level = 1;
    }

    public void upgrade() {
        ++level;
    }

    public void downgrade() {
        Preconditions.checkArgument(level > 1, "You cannot downgrade a level 1 upgrade!");
        --level;
    }

    public int getLevel() {
        return level;
    }

    public abstract void applyUpgradeModifier(MachineResourceGetProcess mineProcess);

    public abstract String getUpgradeName();

    public abstract Map<Integer, Integer> getCosts();

    public int getMaxLevel() {
        if (maxLevel == -1) maxLevel = calculateMaxLevel();
        return maxLevel;
    }

    private int calculateMaxLevel() {
        int max = 1;
        for (int key : getCosts().keySet()) {
            if (key > max) max = key;
        }
        return max;
    }

    public boolean isRunOnlyOnProcessStart() {
        return runOnlyOnProcessStart;
    }

    public abstract UpgradeType getUpgradeType();

    @Override
    public String toString() {
        return "AbstractUpgrade{" +
                "level=" + level +
                '}';
    }
}

File: src/main/kotlin/me/oriharel/machinery/upgrades/AbstractUpgrade.kt
package me.oriharel.machinery.upgrades

import com.google.common.base.Preconditions
import me.oriharel.machinery.Machinery
import me.oriharel.machinery.machine.MachineResourceGetProcess
import org.bukkit.configuration.file.YamlConfiguration

/**
 * abstract implementation of upgrade.
 * abstract since every upgrade has it's own modifying logic but behind the scenes upgrading and downgrading a upgrade is the same.
 */
abstract class AbstractUpgrade {
    var level: Int
        protected set
    var isRunOnlyOnProcessStart: Boolean = false
    var maxLevel: Int? = null
        get() {
            if (field == null) maxLevel = calculateMaxLevel()
            return field
        }

    constructor(level: Int) {
        this.level = level
        isRunOnlyOnProcessStart = false
    }

    protected constructor() {
        level = 1
    }

    fun upgrade() {
        ++level
    }

    fun downgrade() {
        Preconditions.checkArgument(level > 1, "You cannot downgrade a level 1 upgrade!")
        --level
    }

    abstract fun applyUpgradeModifier(mineProcess: MachineResourceGetProcess)
    abstract val upgradeName: String
    abstract val costs: Map<Int, Int>?

    private fun calculateMaxLevel(): Int {
        var max = 1
        for (key in costs!!.keys) {
            if (key > max) max = key
        }
        return max
    }

    abstract val upgradeType: UpgradeType

    override fun toString(): String {
        return "AbstractUpgrade{" +
                "level=" + level +
                '}'
    }

    companion object {
        private var configCache: YamlConfiguration? = null

        val configLoad: YamlConfiguration?
            get() {
                if (configCache != null) {
                    return configCache
                }
                configCache = Machinery.instance?.fileManager?.getConfig("upgrades.yml")?.get();
                return configCache;
            }
    }
}